In [1]:
import numpy as np
import pandas as pd

The dataset is taken from eBay  Kleinanzeigen, a classifieds section of the German eBay website.
The original dataset was scraped and uploaded to Kaggle. We use a symplified version with 50.000 datapoints
and it has been dirtied. The aim of this project is to clean and analyse the included car listing.

In [2]:
autos = pd.read_csv('./DataCleaning/autos.csv', encoding='latin-1')

In [3]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

name is very long,contains additional info, 5 columns('vehicleType', 'gearbox', 'model', 'fuelType', 'notRepairedDamage') contain null values
'price', 'odometer' may be cleaned and converted to number, 'dateCreated' may be converted to date type

In [5]:
def clean_column(col):
    upper = False
    if col == 'yearOfRegistration':
        col = 'registration_year'
    elif col == 'monthOfRegistration':
        col = 'registration_month'
    elif col == 'notRepairedDamage':
        col = 'unrepaired_damage'
    elif col == 'dateCreated':
        col = 'ad_created'
    else:
        col = col.replace(' ', '_')
        new_col = []
        for c in col:
            if c.isupper():
                if not upper:
                    d = '_' + c.lower()
                else:
                    d = c.lower()
                new_col.append(d)
                upper = True
            else:
                new_col.append(c)
                upper = False
        col = ''.join(new_col)
    return col

In [6]:
new_columns = []
columns = autos.columns
for c in columns:
    c = clean_column(c)
    new_columns.append(c)
new_columns
autos.columns = new_columns

Here we have cleaned names of columns. We changed 'CamleCase' to 'snake_case', changed upper register to lower, and changed names of some columns to more readable names.

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [8]:
autos['nr_of_pictures'].unique()

array([0], dtype=int64)

In [9]:
print(autos.describe(include='all'))

               date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
cou

number of pictures column contains only zero and can be dropped

In [10]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

seller contains only 2 unique values, one of them in only one row

In [11]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

offer_type also dosn't contain usefull info, as only in one row the value is different

In [12]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

as we have already seen, we need to clean the columns 'odometer' and 'price'

let's rename 'odometer' to 'odometer_km' as it will show us units when we remove non digits from this column. Remember, we need to do it to be able to filter our data.

In [13]:
autos.rename({'odometer' : 'odometer_km'}, axis=1, inplace=True)

In [14]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','')

C:\Users\Yura\AppData\Local\Temp\ipykernel_2208\1415013928.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos['price'] = autos['price'].str.replace('$','').str.replace(',','')


In [15]:
autos['price'] = autos['price'].astype(int)

In [16]:
autos['odometer_km'] = autos['odometer_km'].str.replace('km','').str.replace(',','')

In [17]:
autos['odometer_km'] = autos['odometer_km'].astype(int)

In [18]:
print(autos['price'].min())
print(autos['price'].max())

0
99999999


In [19]:
expensive_cars = autos['price'] >= 100000
autos.loc[expensive_cars, ['price',  'name', 'registration_year']]

price                                               name  \
514      999999                   Ford_Focus_Turnier_1.6_16V_Style   
1878     129000                                  Porsche_911_Turbo   
2454     137999                                    Porsche_911_GT3   
2751     120000                               Porsche_911___993_4S   
2897   11111111   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7402     115000  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
7814    1300000                                        Ferrari_F40   
8232     128000               Porsche_993_S_Schalter_BRD_neuwertig   
10500    155000                                        Porsche_991   
11137  10000000  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
11433    175000  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
12682    265000  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
14268    135000   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011   
14715    345000                 Rolls_Royce_Phantom_Drophead_Coupe   
16964    105000                    Bentley_Continental_Supersports   
17140    180000  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
17540    105000   Porsche_930_Turbo_classic_analytics_Gutachten_2+   
18509    169000                                    Porsche_991_GT3   
20351    190000                                    BMW_Z8_roadster   
21783    115991               Porsche_911_Carrera_4S_Cabriolet_PDK   
22060    114400  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...   
22673    163991  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
22947   1234566                         Bmw_530d_zum_ausschlachten   
24384  11111111                            Schlachte_Golf_3_gt_tdi   
27371  12345678                                         Fiat_Punto   
28090    194000  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
29286    104900  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...   
32185    139997  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...   
32840    169999  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
33638    151990                                 Porsche_911_GT3_RS   
33884    116000               Porsche_911_Carrera_4S_Cabriolet_PDK   
34723    299000  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923    295000  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
36818    350000                                        Porsche_991   
37585    999990                                Volkswagen_Jetta_GT   
37840    220000                                        Porsche_997   
38299    250000                                Glas_BMW_mit_Wasser   
38814    119500  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...   
39377  12345678                        Tausche_volvo_v40_gegen_van   
39705  99999999                        Tausch_gegen_gleichwertiges   
40918    198000                             Porsche_911_991_GT3_RS   
42221  27322222                                  Leasinguebernahme   
43049    999999                                      2_VW_Busse_T3   
43282    119900               Porsche_911_Carrera_4S_Cabriolet_PDK   
43668    197000  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
44406    120000  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...   
45387    163500  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
47337    259000                                    BMW_Z8_roadster   
47598  12345678  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634   3890000                                        Ferrari_FXX   
49391    109999  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...   
49668    145000         Porsche_911__993__Carrera_S_"SONDERMODELL"   
49815    130000  SUCHE_TIPPS___Ford_Mustang_Shelby_GT_350_500_K...   

       registration_year  
514                 2009  
1878                1995  
2454                2010  
2751                1998  
2897                1973  
7402                2016  
7814                1992  
8232         

In [20]:
cheap_cars = autos['price'].between(500,1000)
autos.loc[cheap_cars, ['price',  'name', 'registration_year']]

price                                       name  registration_year
9        590  Renault_Megane_Scenic_1.6e_RT_Klimaanlage               1997
10       999               VW_Golf_Tuning_in_siber/grau               2017
26       777                        Volkswagen_Polo_Fox               1992
44       900                               Zu_verkaufen               1998
45      1000                         Toyota_Starlet_1.3               1995
...      ...                                        ...                ...
49971    950                                 W.Lupo_1.0               2001
49978    900                Mercedes_Benz_E_200_Classic               1996
49983    600                              Ford_focus_99               1999
49985   1000                  Verkaufe_meinen_vw_vento!               1995
49991    500                                 Kleinwagen               2016

[7110 rows x 3 columns]

In [21]:
autos['price'].unique().shape

(2357,)

In [22]:
autos['price'].value_counts().tail(30)

34550     1
744       1
27600     1
2798      1
55800     1
23350     1
55500     1
22790     1
2944      1
5485      1
23233     1
8098      1
829       1
5248      1
8008      1
5913      1
169000    1
3266      1
4475      1
2035      1
12560     1
15749     1
3945      1
16375     1
410       1
414       1
79933     1
5198      1
18890     1
16995     1
Name: price, dtype: int64

we need to remove rows with prices like 99999, 12345678, 11111111, 27322222 and prices less than 1000, they look unrealistic, for example for focus 99 year - can't be 600 usd,
In my opinion prices under 1000 migh be removed, they look unrealistic

In [23]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [24]:
autos[autos['odometer_km'] >= 140000].head(10)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   

    seller offer_type  price   abtest vehicle_type  registration_year  \
0   privat    Angebot   5000  control          bus               2004   
1   privat    Angebot   8500  control    limousine               1997   
4   privat    Angebot   1350     test        kombi               2003   
5   privat    Angebot   7900     test          bus               2006   
6   privat    Angebot    300     test    limousine               1995   
7   privat    Angebot   1990  control    limousine               1998   
8   privat    Angebot    250     test          NaN               2000   
9   privat    Angebot    590  control          bus               1997   
10  privat    Angebot    999     test          NaN               2017   
11  privat    Angebot    350  control          NaN               2000   

      gearbox  power_ps    model  odometer_km  registration_month fuel_type  \
0     manuell       158   andere       150000                   3       lpg   
1   automatik       286      7er       150000                   6    benzin   
4     manuell         0    focus       150000                   7    benzin   
5   automatik       150  voyager       150000                   4    diesel   
6     manuell        90     golf       150000                   8    benzin   
7     manuell        90     golf       150000                  12    diesel   
8     manuell         0    arosa       150000                  10       NaN   
9     manuell        90   megane       150000                   7    benzin   
10    manuell        90      NaN       150000                   4    benzin   
11        NaN         0      NaN       150000                   0    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
5        chrysler               NaN  2016-03-21 00:00:00               0   
6      volkswagen               NaN  2016-03-20 00:00:00               0   
7      volkswagen              nein  2016-03-16 00:00:00               0   
8            seat              nein  2016-03-22 00:00:00               0   
9         renault              nein  2016-03-16 00:00:00               0   
10     volkswagen              nein  2016-03-14 00:00:00               0   
11  mercedes_benz               NaN  2016-03-16 00:00:00               0   

    postal_code            last_seen  
0         79588  2016-04-06 06:45:54  
1         71034  2016-04-06 14:45:08  
4         39218  2016-04-01 14:38:50  
5         22962  2016-04-06 09:45:21  
6         31535  2016-03-23 02:48:59  
7         53474  2016-04-07 03:17:32  
8          7426  2016-03-26 18:18:10  
9         15749  2016-04-06 10:46:35  
10        86157  2016-04-07 03:16:21  
11        17498  2016-03-16 18:45:34

odometer column looks realistique

In [25]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [26]:
autos = autos[(autos['price'] != 99999999) & (autos['price'] != 12345678) & (autos['price'] != 11111111) & (autos['price'] != 1234566) & (autos['price'] >= 1000)]

In [27]:
autos.loc[expensive_cars, ['price',  'name', 'registration_year']]

price                                               name  \
514      999999                   Ford_Focus_Turnier_1.6_16V_Style   
1878     129000                                  Porsche_911_Turbo   
2454     137999                                    Porsche_911_GT3   
2751     120000                               Porsche_911___993_4S   
7402     115000  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
7814    1300000                                        Ferrari_F40   
8232     128000               Porsche_993_S_Schalter_BRD_neuwertig   
10500    155000                                        Porsche_991   
11137  10000000  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
11433    175000  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
12682    265000  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
14268    135000   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011   
14715    345000                 Rolls_Royce_Phantom_Drophead_Coupe   
16964    105000                    Bentley_Continental_Supersports   
17140    180000  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
17540    105000   Porsche_930_Turbo_classic_analytics_Gutachten_2+   
18509    169000                                    Porsche_991_GT3   
20351    190000                                    BMW_Z8_roadster   
21783    115991               Porsche_911_Carrera_4S_Cabriolet_PDK   
22060    114400  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...   
22673    163991  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
28090    194000  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
29286    104900  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...   
32185    139997  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...   
32840    169999  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
33638    151990                                 Porsche_911_GT3_RS   
33884    116000               Porsche_911_Carrera_4S_Cabriolet_PDK   
34723    299000  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923    295000  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
36818    350000                                        Porsche_991   
37585    999990                                Volkswagen_Jetta_GT   
37840    220000                                        Porsche_997   
38299    250000                                Glas_BMW_mit_Wasser   
38814    119500  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...   
40918    198000                             Porsche_911_991_GT3_RS   
42221  27322222                                  Leasinguebernahme   
43049    999999                                      2_VW_Busse_T3   
43282    119900               Porsche_911_Carrera_4S_Cabriolet_PDK   
43668    197000  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
44406    120000  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...   
45387    163500  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
47337    259000                                    BMW_Z8_roadster   
47634   3890000                                        Ferrari_FXX   
49391    109999  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...   
49668    145000         Porsche_911__993__Carrera_S_"SONDERMODELL"   
49815    130000  SUCHE_TIPPS___Ford_Mustang_Shelby_GT_350_500_K...   

       registration_year  
514                 2009  
1878                1995  
2454                2010  
2751                1998  
7402                2016  
7814                1992  
8232                1997  
10500               2013  
11137               1960  
11433               2016  
12682               2016  
14268               2010  
14715               2012  
16964               2010  
17140               2012  
17540               1988  
18509               2013  
20351               2000  
21783               2013  
22060               2016  
22673               2014  
28090               2016  
29286               2011  
32185               2010  
32840               2012  
33638            

now we have realistic prices for expensive cars

In [28]:
autos['price'].value_counts(ascending=True).head(20)

15100     1
6498      1
3975      1
9845      1
1355      1
20980     1
22199     1
16998     1
1221      1
1689      1
190000    1
35990     1
2498      1
7085      1
2651      1
5049      1
43461     1
7333      1
19970     1
12860     1
Name: price, dtype: int64

In [29]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [30]:
autos['date_crawled'] = autos['date_crawled'].str[:10]

no anomalies is detected

In [31]:
autos['ad_created'] = autos['ad_created'].str[:10]

In [32]:
autos['last_seen'] = autos['last_seen'].str[:10]

we don't need exact time in these columns

In [33]:
autos['registration_year'].describe()

count    38633.00000
mean      2005.67740
std         86.67779
min       1000.00000
25%       2001.00000
50%       2005.00000
75%       2009.00000
max       9999.00000
Name: registration_year, dtype: float64

There can be seen anomaliies as 1000 and 9999, such years look unrealistique

In [34]:
autos[autos['registration_year'].between(1900, 1950)]

date_crawled                                               name  seller  \
2221    2016-03-15  Sehr_seltener_Oldtimer_Opel_1210_zum_Restaurieren  privat   
2573    2016-03-19              Hanomag_rekord_15k_Suche_ersatz_teile  privat   
11047   2016-03-08          Andere_Simca_5_Fourgonette_Kombilimousine  privat   
11246   2016-03-26                  Ford_Model_A_Roadster_Deluxe_1931  privat   
11585   2016-03-11                              Volkswagen__VW_Typ_82  privat   
13963   2016-03-20  Mercedes_Benz_L1500S_Wehrmacht_/_Luftwaffe___F...  privat   
14020   2016-03-19                                 Oldtimeraufloesung  privat   
21416   2016-03-12                            Essex_super_six__Ford_A  privat   
21421   2016-03-05            Ford_Business_Coupe_Hotrod_Projekt.1937  privat   
22101   2016-03-09                                         BMW_Andere  privat   
23804   2016-03-30                                 Mercedes_Benz_170V  privat   
24855   2016-03-31                      Andere_DKW_F8_Cabrio_Limosine  privat   
25792   2016-03-09              Andere_GAZ_Oldtimergelaendewagen_1941  privat   
26103   2016-03-16               Oldtimer_Sport___Cabrio_Georges_Irat  privat   
26607   2016-03-21                                          MG_Andere  privat   
39725   2016-03-25  Dodge_1937_Hotrod_Flathead_Scheunenfund_Oldtim...  privat   

      offer_type  price   abtest vehicle_type  registration_year  gearbox  \
2221     Angebot   3350  control       andere               1934  manuell   
2573     Angebot   3000     test       andere               1934      NaN   
11047    Angebot  17500  control        kombi               1948  manuell   
11246    Angebot  27500  control       cabrio               1931  manuell   
11585    Angebot  41900     test       cabrio               1943      NaN   
13963    Angebot  26900     test       andere               1941  manuell   
14020    Angebot  10000     test        coupe               1950  manuell   
21416    Angebot  16500  control       cabrio               1927  manuell   
21421    Angebot   7000     test        coupe               1937  manuell   
22101    Angebot  11500     test       cabrio               1929  manuell   
23804    Angebot   4800  control    limousine               1937  manuell   
24855    Angebot  26600  control       cabrio               1939      NaN   
25792    Angebot   9500  control          suv               1941  manuell   
26103    Angebot   8500     test       cabrio               1938  manuell   
26607    Angebot  25500  control       cabrio               1937  manuell   
39725    Angebot   8990  control    limousine               1937      NaN   

       power_ps   model  odometer_km  registration_month fuel_type  \
2221          0  andere         5000                   0    benzin   
2573          0     NaN        90000                   1    benzin   
11047         0     NaN        60000                   6    benzin   
11246        39  andere        10000                   7    benzin   
11585         0  andere       100000                   7       NaN   
13963        60  andere        60000                   7    benzin   
14020       130  andere         5000                   1    benzin   
21416        40  andere         5000                   5    benzin   
21421        85  andere         5000                   8    benzin   
22101        15  andere         5000                   1       NaN   
23804        38  andere         5000                   0    benzin   
24855         0     NaN         5000                   7       NaN   
25792        57     NaN        40000                   1    benzin   
26103         0     NaN         5000                   6    benzin   
26607        50     NaN        50000                   5    benzin   
39725         0     NaN       150000                   7    benzin   

                brand unrepaired_damage  ad_created  nr_of_pictures  \
2221             opel                ja  2016-03-15     

dates between 1900 and 1950 look realistique

In [35]:
autos[autos['registration_year'] <= 1900]

date_crawled                                   name  seller offer_type  \
22316   2016-03-29  VW_Kaefer.__Zwei_zum_Preis_von_einem.  privat    Angebot   
49283   2016-03-15                             Citroen_HY  privat    Angebot   

       price   abtest vehicle_type  registration_year  gearbox  power_ps  \
22316   1500  control          NaN               1000  manuell         0   
49283   7750  control          NaN               1001      NaN         0   

        model  odometer_km  registration_month fuel_type       brand  \
22316  kaefer         5000                   0    benzin  volkswagen   
49283  andere         5000                   0       NaN     citroen   

      unrepaired_damage  ad_created  nr_of_pictures  postal_code   last_seen  
22316               NaN  2016-03-29               0        48324  2016-03-31  
49283               NaN  2016-03-15               0        66706  2016-04-06

these should be removed as well as all the rows with registration year > 2016, as it can't be higher that year for crawling process

In [36]:
autos[autos['registration_year'] > 2016]

date_crawled                                               name  seller  \
68      2016-04-03  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...  privat   
113     2016-04-03                              Golf_4_Anfaenger_auto  privat   
164     2016-03-13  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...  privat   
253     2016-03-27             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017  privat   
348     2016-03-17  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...  privat   
...            ...                                                ...     ...   
49689   2016-03-14                                      VW_GOLF_V_1.9  privat   
49696   2016-03-08               Audi_A3_1_6_Attraction_mit_TÜV_10/17  privat   
49731   2016-03-28  Renault_MEGANE_1.6_16V__neu_ZANHRIEMEN_TOP_ZUS...  privat   
49796   2016-03-09                        Opel_corsa_1.4_zu_verkaufen  privat   
49910   2016-04-03  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...  privat   

      offer_type  price   abtest vehicle_type  registration_year    gearbox  \
68       Angebot  10990     test          NaN               2017    manuell   
113      Angebot   1200     test          NaN               2017    manuell   
164      Angebot   4800  control          NaN               2018    manuell   
253      Angebot   2250     test          NaN               2017    manuell   
348      Angebot   3750  control          NaN               2017    manuell   
...          ...    ...      ...          ...                ...        ...   
49689    Angebot   4700  control          NaN               2017    manuell   
49696    Angebot   1250  control          NaN               2017    manuell   
49731    Angebot   1700     test          NaN               2017    manuell   
49796    Angebot   4500     test          NaN               2017    manuell   
49910    Angebot  22200     test          NaN               9000  automatik   

       power_ps    model  odometer_km  registration_month fuel_type  \
68          174  clubman       100000                   0       NaN   
113          75     golf       150000                   7       NaN   
164           0   meriva        80000                   4    benzin   
253           0   mondeo       150000                   8    benzin   
348         150   beetle       150000                   7       NaN   
...         ...      ...          ...                 ...       ...   
49689       105      NaN       150000                   0    diesel   
49696       101       a3       150000                   8    benzin   
49731         0   megane       150000                   7       NaN   
49796        90    corsa        70000                   7    benzin   
49910       140   andere        10000                   3    benzin   

            brand unrepaired_damage  ad_created  nr_of_pictures  postal_code  \
68           mini              nein  2016-04-03               0        83135   
113    volkswagen               NaN  2016-04-03               0        97656   
164          opel              nein  2016-03-13               0        37627   
253          ford              nein  2016-03-27               0        56575   
348    volkswagen              nein  2016-03-17               0        45896   
...           ...               ...         ...             ...          ...   
49689  volkswagen               NaN  2016-03-14               0        22459   
49696        audi              nein  2016-03-08               0        26624   
49731     renault              nein  2016-03-28               0        34123   
49796        opel              nein  2016-03-09               0        88433   
49910        opel               NaN  2016-04-03               0        56856   

        last_seen  
68     2016-04-05  
113    2016-04-05  
164    2016-04-04  
253    2016-04-05  
348    2016-03-24  
...           ...  
49689  2016-03-16  
49696  2016-03-08  
49731  2016-03-31  
49796  2016-03-17  
49910  2016-04-05  

[1418 rows x 20 columns]

In [37]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [38]:
autos['registration_year'].value_counts(normalize=True)

2005    0.074839
2006    0.071265
2004    0.070083
2003    0.066563
2007    0.060705
          ...   
1939    0.000027
1948    0.000027
1938    0.000027
1953    0.000027
1950    0.000027
Name: registration_year, Length: 77, dtype: float64

The most rare models are build before 1950

In [39]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'audi', 'renault', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'jeep', 'saab',
       'volvo', 'nissan', 'jaguar', 'skoda', 'subaru', 'fiat',
       'mitsubishi', 'chevrolet', 'hyundai', 'honda', 'kia', 'citroen',
       'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'rover',
       'daihatsu', 'daewoo', 'lancia', 'lada'], dtype=object)

In [40]:
autos['brand'].value_counts()

volkswagen        7845
bmw               4664
mercedes_benz     4152
audi              3631
opel              3314
ford              2185
renault           1387
peugeot           1038
fiat               784
skoda              709
seat               643
smart              618
toyota             544
mazda              530
citroen            518
nissan             507
mini               405
hyundai            400
sonstige_autos     392
volvo              334
kia                286
porsche            278
honda              273
mitsubishi         256
chevrolet          246
alfa_romeo         232
suzuki             213
dacia              122
chrysler           118
jeep               103
land_rover          98
jaguar              69
subaru              64
daihatsu            63
saab                51
daewoo              34
trabant             32
rover               27
lancia              25
lada                23
Name: brand, dtype: int64

In [41]:
autos[autos['brand'] == 'sonstige_autos']

date_crawled                                               name  seller  \
15      2016-04-01        Corvette_C3_Coupe_T_Top_Crossfire_Injection  privat   
140     2016-03-19           Ssangyong_Actyon_SUV_2.0_xdi2wd_55000_km  privat   
152     2016-03-20                                Ssanyong_Rexton_2.7  privat   
175     2016-03-19                                          MG_MGB_GT  privat   
296     2016-03-07                      Dodge_Nitro_4.0_Automatik_R/T  privat   
...            ...                                                ...     ...   
49271   2016-03-16                                        Corvette_C6  privat   
49391   2016-03-18  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...  privat   
49720   2016-03-31                                          MG_Andere  privat   
49745   2016-03-19                                         ALPINA_B12  privat   
49941   2016-03-15                   Maserati_Ghibli_Diesel_Automatik  privat   

      offer_type   price   abtest vehicle_type  registration_year    gearbox  \
15       Angebot   18900     test        coupe               1982  automatik   
140      Angebot    5400  control          suv               2008    manuell   
152      Angebot    4499     test          suv               2005  automatik   
175      Angebot   13800  control        coupe               1972        NaN   
296      Angebot   11500  control          suv               2007  automatik   
...          ...     ...      ...          ...                ...        ...   
49271    Angebot   28900     test       cabrio               2005  automatik   
49391    Angebot  109999     test        coupe               2008  automatik   
49720    Angebot   14500     test        coupe               1969    manuell   
49745    Angebot   17500     test    limousine               1996        NaN   
49941    Angebot   62000     test    limousine               2015  automatik   

       power_ps model  odometer_km  registration_month fuel_type  \
15          203   NaN        80000                   6    benzin   
140         141   NaN        60000                   3    diesel   
152         163   NaN       150000                   9    diesel   
175           0   NaN        20000                   6       NaN   
296         260   NaN       100000                   7    benzin   
...         ...   ...          ...                 ...       ...   
49271       404   NaN        90000                   7    benzin   
49391       560   NaN        30000                   9    benzin   
49720       145   NaN        50000                   1    benzin   
49745       387   NaN       150000                   5       NaN   
49941       275   NaN        10000                   8    diesel   

                brand unrepaired_damage  ad_created  nr_of_pictures  \
15     sonstige_autos              nein  2016-04-01               0   
140    sonstige_autos              nein  2016-03-19               0   
152    sonstige_autos              nein  2016-03-20               0   
175    sonstige_autos              nein  2016-03-19               0   
296    sonstige_autos              nein  2016-03-07               0   
...               ...               ...         ...             ...   
49271  sonstige_autos              nein  2016-03-16               0   
49391  sonstige_autos              nein  2016-03-17               0   
49720  sonstige_autos              nein  2016-03-31               0   
49745  sonstige_autos              nein  2016-03-19               0   
49941  sonstige_autos              nein  2016-03-14               0   

       postal_code   last_seen  
15           61276  2016-04-02  
140          94447  2016-03-26  
152          73312  2016-03-25  
175          53639  2016-03-23  
296          88212  2016-04-07  
...            ...         ...  
49271        34121  2016-04-06  
49391        96052  2016-04-05  
49720        86911  2016-04-06  
49745        76131  2016-03-20  
49941        10179  2016-03-30  

[392 

Brand 'sonstige_autos' looks abnormal, we may try to get the brand name from name, part before first underscore

In [42]:
brand_error = autos['brand'] == 'sonstige_autos'

We add additional column 's_brand' and if there is no brand, we extract first part of the name and use it brand but keep original column intact

In [43]:
autos['s_brand'] = autos[brand_error]['name'].str.extract(r'^([A-Za-z]{2,})_.*?')

In [44]:
autos.loc[brand_error, 's_brand'].unique()

array(['Corvette', 'Ssangyong', 'Ssanyong', 'MG', 'Dodge', 'Werkaufen',
       'Pontiac', 'Cadillac', 'Andere', 'Buick', 'Wartburg', nan, 'DKW',
       'VW', 'MICROCAR', 'Verkaufe', 'Super', 'Abarth', 'Hanomag',
       'Ferrari', 'Iveco', 'Maserati', 'Melkus', 'Triumph', 'Alpina',
       'Suche', 'Pfau', 'NSU', 'GMC', 'Combi', 'Strandbuggy', 'Talbot',
       'Elektroauto', 'AC', 'Borgward', 'Lexus', 'Mopedauto', 'Gmc',
       'Bentley', 'Plymouth', 'Barkas', 'Asia', 'Oldsmobil', 'Sehr',
       'MOPEDAUTO', 'LEXUS', 'suche', 'MGF', 'Isuzu', 'Lincoln',
       'Wohnwagen', 'Chevrolet', 'Beach', 'Aixam', 'Rolls', 'ALPINA',
       'Piaggio', 'Vw', 'Morgan', 'Hummer', 'FREEWIEL', 'maserati',
       'Saangyong', 'Microcar', 'Massey', 'Biete', 'Tesla', 'Oldtimer',
       'Morris', 'Willys', 'Aston', 'Wir', 'Spitfire', 'Zu', 'MASERATI',
       'Buggy', 'ML', 'Oldsmobile', 'Suesser', 'Brilliance', 'GS',
       'Lotus', 'Robur', 'Ford', 'Reno', 'MAN', 'ssangyoug', 'Zastava',
       'Westfield', '

In [45]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'audi', 'renault', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'jeep', 'saab',
       'volvo', 'nissan', 'jaguar', 'skoda', 'subaru', 'fiat',
       'mitsubishi', 'chevrolet', 'hyundai', 'honda', 'kia', 'citroen',
       'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'rover',
       'daihatsu', 'daewoo', 'lancia', 'lada'], dtype=object)

ssnagyong has three different description variants, we may remove rows with brands ('*', '25', '1936', 'Sehr',  'Zu', '1960', 'Strandbuggy', 'Ape',  'Unfallwagen', NaN, 'LKW', '2er', 'suesser', '1969er', 'Auto'), 

In [46]:
autos = autos[~autos['s_brand'].isin(['Andere', 'Werkaufen', 'super', 'Suche',  'Pfau',  'Combi',  'Strandbuggy',  'Elektroauto',  'AC',  'Sehr',  'Wohnwagen',  'Beach',  'FREEWIEL',  'Oldtimer',  'Wir',  'Zu',  'Suesser',  'Brilliance',  'GS', 'suche', 'Westfield',  'Ape',  'Oldtimersammlung',  'Auto',  'Unfallwagen'])]

check if we have Nan in 's_brand' and 'sonstige_autos' in brand in one row

In [47]:
autos[(autos['brand'] == 'sonstige_autos') & (autos['s_brand'].isnull())]

date_crawled                                               name  seller  \
1548    2016-03-25               1936_Daimler_Fifteen_Special_Projekt  privat   
7294    2016-03-28      YA126_DAF_Berge__Werstatt_und_Manschaft_Wagen  privat   
11817   2016-04-01  Strandbuggy/_Borossi_Joyner_BB_1100_Strassenzu...  privat   
18315   2016-03-08                      2er_pferdeanhaenger_boeckmann  privat   
25500   2016-03-21                      25_Km_Auto_ohne_Fuehrerschein  privat   
25683   2016-03-08                                 SportSLTQuadCab4x4  privat   
28717   2016-03-26  1960_Nash_Metropolitan_Rockabella_Dream_Car_mi...  privat   
33650   2016-03-12  *_Ssyangyong_Rodius_270_XDI_aus_erste_hand_*_A...  privat   
36621   2016-04-01                            V8_Olds_mit_H_Zulassung  privat   
46418   2016-04-02  Steyr/Puch_Haflinger_700_AP_4x4_BJ_1971_Milita...  privat   
47445   2016-03-31             1969er_Dodge_D200_Pickup_Autom._383cui  privat   

      offer_type  price   abtest vehicle_type  registration_year    gearbox  \
1548     Angebot   7999  control          NaN               1960        NaN   
7294     Angebot   9990     test          suv               1956    manuell   
11817    Angebot   8500     test       cabrio               2009    manuell   
18315    Angebot   2900  control          suv               1994    manuell   
25500    Angebot   1500     test   kleinwagen               2000        NaN   
25683    Angebot   8555     test          suv               2009  automatik   
28717    Angebot   9900  control        coupe               1960    manuell   
33650    Angebot   1999     test          suv               2006  automatik   
36621    Angebot   7999     test    limousine               1985  automatik   
46418    Angebot   8500  control          suv               1971    manuell   
47445    Angebot   7300     test          suv               1969  automatik   

       power_ps  ... odometer_km  registration_month  fuel_type  \
1548         65  ...      150000                   6     benzin   
7294         75  ...       20000                   6     benzin   
11817        68  ...       20000                   6     benzin   
18315         0  ...       10000                   4     diesel   
25500        20  ...      100000                   0        NaN   
25683       345  ...      100000                   8     benzin   
28717        50  ...       70000                   7     benzin   
33650       170  ...      150000                  11     diesel   
36621       150  ...       60000                   7     benzin   
46418        27  ...       20000                   5     benzin   
47445         0  ...      100000                   1     benzin   

                brand unrepaired_damage  ad_created nr_of_pictures  \
1548   sonstige_autos               NaN  2016-03-25              0   
7294   sonstige_autos              nein  2016-03-28              0   
11817  sonstige_autos              nein  2016-04-01              0   
18315  sonstige_autos              nein  2016-03-08              0   
25500  sonstige_autos               NaN  2016-03-21              0   
25683  sonstige_autos              nein  2016-03-08              0   
28717  sonstige_autos              nein  2016-03-26              0   
33650  sonstige_autos              nein  2016-03-12              0   
36621  sonstige_autos               NaN  2016-04-01              0   
46418  sonstige_autos               NaN  2016-04-02              0   
47445  sonstige_autos              nein  2016-03-31              0   

       postal_code   last_seen s_brand  
1548         50259  2016-04-06     NaN  
7294         56206  2016-04-07     NaN  
11817        66780  2016-04-05     NaN  
18315        48455  2016-03-09     NaN  
25500        44894  2016-04-05     NaN  
25683        77933  2016-03-08     NaN  
28717        17268  2016-04-06     NaN  
33650        13469  2016-03-17     NaN  
36621        28207  2016-04-05     NaN  
46418        56379  2016-04-

only 11 row, we may drop them as their name doesn't follow pattern we have used and without brand name we won't be able to make statistics

In [48]:
remove_condition = (autos['brand'] == 'sonstige_autos') & (autos['s_brand'].isnull())
autos = autos[~remove_condition]

In [49]:
autos['s_brand'].fillna(autos['brand'], inplace=True)

In [50]:
autos[autos['s_brand'].isnull()]

Empty DataFrame
Columns: [date_crawled, name, seller, offer_type, price, abtest, vehicle_type, registration_year, gearbox, power_ps, model, odometer_km, registration_month, fuel_type, brand, unrepaired_damage, ad_created, nr_of_pictures, postal_code, last_seen, s_brand]
Index: []

[0 rows x 21 columns]

'Corvette' change to 'Chevrolet', 'vw' to 'volkswagen', all valiation of 'Ssangyong' should be replaced by the latter, 'oldtimer' must be removed, and all values in 's_brand' should be in lowercase

In [51]:
vw_bool = autos['s_brand'] == 'vw'
autos.loc[vw_bool, 's_brand'] = 'volkswagen'

In [52]:
autos.loc[autos['s_brand'] == 'Corvette', 's_brand'] = 'chevrolet'
autos.loc[autos['s_brand'] == 'Ssanyong', 's_brand'] = 'ssangyong'
autos.loc[autos['s_brand'] == 'Ssanyong', 's_brand'] = 'ssangyong'
autos.loc[autos['s_brand'] == 'Saangyong', 's_brand'] = 'ssangyong'
autos.loc[autos['s_brand'] == 'combi', 's_brand'] = 'asia'
autos.loc[autos['s_brand'] == 'reno', 's_brand'] = 'renault'
autos.loc[autos['s_brand'] == 'mini', 's_brand'] = 'bmw'
autos.loc[autos['s_brand'] == 'Alpina', 's_brand'] = 'bmw'

In [53]:
autos['s_brand'] = autos['s_brand'].str.lower()

now let's make some analysis

we want to see mean price by brand

In [54]:
brands_mean = {}
brands = list(autos['s_brand'].unique())

In [55]:
brands

['peugeot',
 'bmw',
 'volkswagen',
 'smart',
 'ford',
 'chrysler',
 'audi',
 'renault',
 'chevrolet',
 'mazda',
 'porsche',
 'mercedes_benz',
 'seat',
 'toyota',
 'opel',
 'dacia',
 'jeep',
 'saab',
 'volvo',
 'nissan',
 'jaguar',
 'skoda',
 'subaru',
 'ssangyong',
 'fiat',
 'mg',
 'mitsubishi',
 'hyundai',
 'honda',
 'kia',
 'citroen',
 'dodge',
 'suzuki',
 'trabant',
 'pontiac',
 'land_rover',
 'alfa_romeo',
 'cadillac',
 'rover',
 'daihatsu',
 'buick',
 'wartburg',
 'daewoo',
 'dkw',
 'vw',
 'microcar',
 'verkaufe',
 'super',
 'abarth',
 'hanomag',
 'ferrari',
 'iveco',
 'maserati',
 'melkus',
 'triumph',
 'lancia',
 'nsu',
 'lada',
 'gmc',
 'talbot',
 'borgward',
 'lexus',
 'mopedauto',
 'bentley',
 'plymouth',
 'barkas',
 'asia',
 'oldsmobil',
 'mgf',
 'isuzu',
 'lincoln',
 'aixam',
 'rolls',
 'alpina',
 'piaggio',
 'morgan',
 'hummer',
 'massey',
 'biete',
 'tesla',
 'morris',
 'willys',
 'aston',
 'spitfire',
 'buggy',
 'ml',
 'oldsmobile',
 'lotus',
 'robur',
 'reno',
 'man',
 

In [56]:
num_cars = {}
means = {}
totals = {}
for brand in brands:
    selected_rows = autos[autos['s_brand'] == brand]
    mean = selected_rows['price'].mean()
    total = selected_rows['price'].sum()
    num = selected_rows['s_brand'].count()
    num_cars[brand] = num
    means[brand] = mean
    totals[brand] = total

We iterate list of unique brands, get all rows corresponding to the brand, get statistic and add it to the dictionary

In [57]:
mean_df = pd.DataFrame.from_dict(data=means, orient='index', columns=['mean'] )
totals_df = pd.DataFrame.from_dict(data=totals, orient='index', columns=['total'] )
num_cars = pd.DataFrame.from_dict(data=num_cars, orient='index', columns=['number_ads'] )

We have created dataframes from dictionaries with statistics

In [58]:
mean_df[mean_df['mean']==mean_df['mean'].max()]

mean
ferrari  1326087.5

In [59]:
mean_df[mean_df['mean']==mean_df['mean'].min()]

mean
mgf   1250.0
reno  1250.0

The most expensive average price for car - is for Ferrari, the least expensive is 'mgf'

In [60]:
mean_df.sort_values(by='mean', inplace=True, ascending=False)

In [61]:
mean_df

mean
ferrari      1.326088e+06
tesla        1.293000e+05
lamborghini  1.099990e+05
rolls        1.087250e+05
melkus       8.000000e+04
...                   ...
liger        1.800000e+03
daewoo       1.539559e+03
massey       1.300000e+03
mgf          1.250000e+03
reno         1.250000e+03

[101 rows x 1 columns]

The most expensive cars are - 'ferrari', 'tesla', 'lamborghini', 'rolls', 'melkus'
The least expensive are - 'liger', , 'daewoo', 'massey', 'mgf'

but let's check the brand wich have more than ten ads, because 'ferrari' looks like outlier

In [62]:
frequent_brands = num_cars[num_cars['number_ads'] > 10]

In [63]:
print(frequent_brands.head())
print(frequent_brands.tail())

            number_ads
peugeot           1038
bmw               5070
volkswagen        7845
smart              618
ford              2186
          number_ads
maserati          14
triumph           12
lancia            25
lada              23
lexus             21


iterate through list of brands that we have created, select rows from autos and create new frequency table for mean

In [64]:
frequent_means = {}
for brand in brands:
    selected_rows = autos[autos['s_brand'] == brand]
    mean = selected_rows['price'].mean()
    count = selected_rows['s_brand'].count()
    if count > 10:
        frequent_means[brand] = mean

In [65]:
frequent_means_df = pd.DataFrame.from_dict(data=frequent_means, orient='index', columns=['mean'] )

In [66]:
frequent_means_df[frequent_means_df['mean']==frequent_means_df['mean'].max()]

mean
citroen  57351.664093

In [67]:
frequent_means_df.sort_values(by='mean', inplace=True, ascending=False)

citroen looks too expensive, let's get all rows and check what is wrong

In [68]:
citroens = autos[autos['s_brand'] == 'citroen']

In [69]:
citroens

date_crawled                                               name  seller  \
287     2016-03-26               Citroën_Berlingo_1.6_HDi_110_FAP_XTR  privat   
507     2016-03-08                                      Citroën_Xsara  privat   
884     2016-04-04  Citroen_C1_bj_2009_Tuev_03/2018_Inspektion_Ölw...  privat   
932     2016-03-25                                     Hallo_verkaufe  privat   
983     2016-03-16                   Citroën_Berlingo_Kastenwagen_LKW  privat   
...            ...                                                ...     ...   
49768   2016-03-22  Citroën_C3_Pluriel1.4_Exclusive_SCHECKHEFT_ZAH...  privat   
49885   2016-03-29                         Citroën_Xsara_Picasso_1.8i  privat   
49916   2016-03-28      Citroën_C4_Picasso_2.0_HDi_FAP_ESG6_Exclusive  privat   
49966   2016-04-02  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...  privat   
49970   2016-03-21  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...  privat   

      offer_type  price   abtest vehicle_type  registration_year    gearbox  \
287      Angebot   9850  control          bus               2011    manuell   
507      Angebot   1250  control          NaN               2016    manuell   
884      Angebot   3899     test   kleinwagen               2009    manuell   
932      Angebot   1400     test          NaN               2000        NaN   
983      Angebot   7900     test       andere               2010        NaN   
...          ...    ...      ...          ...                ...        ...   
49768    Angebot   3390  control       cabrio               2006    manuell   
49885    Angebot   1300     test          bus               2000    manuell   
49916    Angebot   6500  control        kombi               2009  automatik   
49966    Angebot   1490  control   kleinwagen               2006    manuell   
49970    Angebot  15800  control          bus               2010  automatik   

       power_ps  ... odometer_km  registration_month  fuel_type    brand  \
287         111  ...       60000                  11     diesel  citroen   
507         109  ...      150000                   7     benzin  citroen   
884          68  ...       60000                   3     benzin  citroen   
932           0  ...      150000                   0        NaN  citroen   
983          75  ...       70000                   0     diesel  citroen   
...         ...  ...         ...                 ...        ...      ...   
49768        73  ...      150000                   7     benzin  citroen   
49885       111  ...      150000                   7     benzin  citroen   
49916       100  ...      150000                   3     diesel  citroen   
49966        68  ...      150000                   7     benzin  citroen   
49970       136  ...       60000                   4     diesel  citroen   

      unrepaired_damage  ad_created nr_of_pictures  postal_code   last_seen  \
287                nein  2016-03-26              0        53894  2016-03-26   
507                nein  2016-03-08              0        23883  2016-03-09   
884                nein  2016-04-04              0        66333  2016-04-06   
932                 NaN  2016-03-25              0        66909  2016-04-06   
983                 NaN  2016-03-16              0         6449  2016-03-21   
...                 ...         ...            ...          ...         ...   
49768              nein  2016-03-22              0        88348  2016-04-05   
49885              nein  2016-03-29              0        65462  2016-03-30   
49916              nein  2016-03-28              0        82538  2016-04-06   
49966                ja  2016-04-02              0        26603  2016-04-02   
49970              nein  2016-03-21              0        14947  2016-04-07   

       s_brand  
287    citroen  
507    citroen  
884    citroen  
932    citroen  
983    citroen  
...        ...  
49768  citroen  
49885  citroen  
49916  citroen  
49966  citroen  
49970  citroen  

[518 rows x 21 co

let's sort by price descending and get top 5 expensive cars

In [70]:
citroens = citroens.sort_values(by='price', ascending=False)

In [71]:
citroens

date_crawled                                               name  seller  \
42221   2016-03-08                                  Leasinguebernahme  privat   
48944   2016-03-21            Citroën_C2_1.6_16V_Senso_Drive_VTR_Plus  privat   
6476    2016-03-29  Citroen_DS20_Pallas_1973_Halbautomatik_/_wenig...  privat   
21720   2016-03-30       Citroën_C5_Tourer_HDi_200_FAP_Aut._Exclusive  privat   
11315   2016-04-01      Roadster__2CV6__Unikat__Karosserie_kompl._Alu  privat   
...            ...                                                ...     ...   
46878   2016-04-05                              Citroen_xsara_picasso  privat   
20504   2016-03-11                             Citroen_Saxo_1.1_Tonic  privat   
19871   2016-04-03                           Citroën_c3_mit_neuem_TÜV  privat   
18726   2016-03-08                                    Citroen_Picasso  privat   
5467    2016-03-19                     verkaufe_Citroen_xsara_Picasso  privat   

      offer_type     price   abtest vehicle_type  registration_year  \
42221    Angebot  27322222  control    limousine               2014   
48944    Angebot     33000  control   kleinwagen               2005   
6476     Angebot     25000  control          NaN               1973   
21720    Angebot     22900  control        kombi               2013   
11315    Angebot     21111     test       cabrio               1986   
...          ...       ...      ...          ...                ...   
46878    Angebot      1000     test          bus               2000   
20504    Angebot      1000  control          NaN               2003   
19871    Angebot      1000     test   kleinwagen               2004   
18726    Angebot      1000  control          NaN               2005   
5467     Angebot      1000     test          bus               2000   

         gearbox  power_ps  ... odometer_km  registration_month  fuel_type  \
42221    manuell       163  ...       40000                   2     diesel   
48944  automatik       109  ...      100000                   1     benzin   
6476         NaN         0  ...      125000                   0        NaN   
21720  automatik       204  ...       60000                   4     diesel   
11315    manuell        27  ...       10000                   0     benzin   
...          ...       ...  ...         ...                 ...        ...   
46878    manuell         0  ...      150000                   3     benzin   
20504    manuell         0  ...       60000                   8        NaN   
19871  automatik        73  ...      150000                   0     benzin   
18726    manuell         0  ...      150000                   0     benzin   
5467     manuell         0  ...      150000                   4     benzin   

         brand unrepaired_damage  ad_created nr_of_pictures  postal_code  \
42221  citroen               NaN  2016-03-08              0        76532   
48944  citroen              nein  2016-03-21              0        21077   
6476   citroen               NaN  2016-03-29              0        67489   
21720  citroen              nein  2016-03-30              0        46147   
11315  citroen               NaN  2016-04-01              0        53175   
...        ...               ...         ...            ...          ...   
46878  citroen               NaN  2016-04-05              0        16816   
20504  citroen              nein  2016-03-11              0         8066   
19871  citroen               NaN  2016-04-03              0        51570   
18726  citroen               NaN  2016-03-08              0        26721   
5467   citroen              nein  2016-03-19              0         9306   

        last_seen  s_brand  
42221  2016-03-08  citroen  
48944  2016-04-06  citroen  
6476   2016-04-06  citroen  
21720  2016-04-05  citroen  
11315  2016-04-05  citroen  
...           ...      ...  
46878  2016-04-05  citroen  
20504  2016-03-19  citroen  
19871  2016-04-03  citroen  
18726  2016-03-10  citroen  
5467   2016

let's remove row with 'Leasinguebernahme'

In [72]:
autos = autos[autos['name'] != 'Leasinguebernahme']

In [73]:
frequent_means = {}
for brand in brands:
    selected_rows = autos[autos['s_brand'] == brand]
    mean = selected_rows['price'].mean()
    count = selected_rows['s_brand'].count()
    if count > 10:
        frequent_means[brand] = mean
frequent_means_df = pd.DataFrame.from_dict(data=frequent_means, orient='index', columns=['mean'] )
frequent_means_df.sort_values(by='mean', inplace=True, ascending=False)

In [74]:
frequent_means_df

mean
porsche        46955.151079
maserati       27719.214286
land_rover     19108.091837
triumph        16438.250000
dodge          13099.755556
jaguar         12295.971014
pontiac        12241.583333
cadillac       12105.909091
jeep           11965.563107
audi           10322.269347
lexus           9816.142857
mercedes_benz   9302.614403
bmw             9248.348915
chevrolet       7814.620155
ssangyong       7393.666667
volkswagen      6898.376546
skoda           6836.696756
mg              6829.730769
kia             6810.160839
nissan          6428.428008
hyundai         6181.872500
volvo           6151.583832
lancia          6135.480000
dacia           5959.098361
subaru          5858.203125
ford            5786.343550
seat            5638.640747
toyota          5573.571691
mazda           5309.526415
honda           5297.282051
wartburg        5264.916667
alfa_romeo      5255.379310
suzuki          5149.877934
mitsubishi      4812.507812
citroen         4614.970986
chrysler        4560.076271
saab            4528.980392
opel            4219.954737
fiat            4008.174745
peugeot         3955.169557
smart           3780.469256
renault         3590.942322
trabant         3131.937500
lada            3032.391304
rover           2862.555556
daihatsu        2546.761905
daewoo          1539.558824

let's analyse top 20 brand by number of sales

In [75]:
top_20_brands = autos['s_brand'].value_counts().head(20).index.tolist()

we get value counts for brands and used first 20 rows, we added to list names in the index column

In [76]:
top_20_brands

['volkswagen',
 'bmw',
 'mercedes_benz',
 'audi',
 'opel',
 'ford',
 'renault',
 'peugeot',
 'fiat',
 'skoda',
 'seat',
 'smart',
 'toyota',
 'mazda',
 'citroen',
 'nissan',
 'hyundai',
 'volvo',
 'kia',
 'porsche']

We will analyse top 6 brands

In [77]:
top_6_brands = ['volkswagen',
 'bmw',
 'mercedes_benz',
 'audi',
 'opel',
 'ford']
frequent_means_top_6 = {}
for brand in top_6_brands:
    selected_rows = autos[autos['s_brand'] == brand]
    mean = selected_rows['price'].mean()
    count = selected_rows['s_brand'].count()
    frequent_means_top_6[brand] = mean
        


In [78]:
frequent_means_df_top_6 = pd.DataFrame.from_dict(data=frequent_means_top_6, orient='index', columns=['mean'] )
frequent_means_df_top_6.sort_values(by='mean', inplace=True, ascending=False)

In [79]:
frequent_means_df_top_6

mean
audi           10322.269347
mercedes_benz   9302.614403
bmw             9248.348915
volkswagen      6898.376546
ford            5786.343550
opel            4219.954737

we will look into possible correlation between avergae mileage and average price for top 6 brands

In [80]:
avg_mileage = {}
for brand in top_6_brands:
    selected_rows = autos[autos['s_brand'] == brand]
    mean = selected_rows['odometer_km'].mean()
    avg_mileage[brand] = mean

In [81]:
avg_mileage

{'volkswagen': 125771.82919056724,
 'bmw': 128534.516765286,
 'mercedes_benz': 130062.6204238921,
 'audi': 127491.04929771413,
 'opel': 123952.92697646348,
 'ford': 119636.32204940531}

In [82]:
mileage_series = pd.Series(avg_mileage)

In [83]:
frequent_means_df_top_6['mean_mileage'] = mileage_series

In [84]:
frequent_means_df_top_6

mean   mean_mileage
audi           10322.269347  127491.049298
mercedes_benz   9302.614403  130062.620424
bmw             9248.348915  128534.516765
volkswagen      6898.376546  125771.829191
ford            5786.343550  119636.322049
opel            4219.954737  123952.926976

There is no vivible correlation between average price and average mileage

We have german words as categorical value, it is advised to chagle them to english euquivalents

In [85]:
autos.head(10)

date_crawled                                               name  seller  \
0    2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1    2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2    2016-03-26                         Volkswagen_Golf_1.6_United  privat   
3    2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4    2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   
5    2016-03-21  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...  privat   
7    2016-03-16                               Golf_IV_1.9_TDI_90PS  privat   
12   2016-03-31  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...  privat   
13   2016-03-23                                 Audi_A3_1.6_tuning  privat   
14   2016-03-23  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...  privat   

   offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0     Angebot   5000  control          bus               2004    manuell   
1     Angebot   8500  control    limousine               1997  automatik   
2     Angebot   8990     test    limousine               2009    manuell   
3     Angebot   4350  control   kleinwagen               2007  automatik   
4     Angebot   1350     test        kombi               2003    manuell   
5     Angebot   7900     test          bus               2006  automatik   
7     Angebot   1990  control    limousine               1998    manuell   
12    Angebot   5299  control   kleinwagen               2010  automatik   
13    Angebot   1350  control    limousine               1999    manuell   
14    Angebot   3999     test   kleinwagen               2007    manuell   

    power_ps  ... odometer_km  registration_month  fuel_type       brand  \
0        158  ...      150000                   3        lpg     peugeot   
1        286  ...      150000                   6     benzin         bmw   
2        102  ...       70000                   7     benzin  volkswagen   
3         71  ...       70000                   6     benzin       smart   
4          0  ...      150000                   7     benzin        ford   
5        150  ...      150000                   4     diesel    chrysler   
7         90  ...      150000                  12     diesel  volkswagen   
12        71  ...       50000                   9     benzin       smart   
13       101  ...      150000                  11     benzin        audi   
14        75  ...      150000                   9     benzin     renault   

   unrepaired_damage  ad_created nr_of_pictures  postal_code   last_seen  \
0               nein  2016-03-26              0        79588  2016-04-06   
1               nein  2016-04-04              0        71034  2016-04-06   
2               nein  2016-03-26              0        35394  2016-04-06   
3               nein  2016-03-12              0        33729  2016-03-15   
4               nein  2016-04-01              0        39218  2016-04-01   
5                NaN  2016-03-21              0        22962  2016-04-06   
7               nein  2016-03-16              0        53474  2016-04-07   
12              nein  2016-03-31              0        34590  2016-04-06   
13              nein  2016-03-23              0        12043  2016-04-01   
14               NaN  2016-03-23              0        81737  2016-04-01   

       s_brand  
0      peugeot  
1          bmw  
2   volkswagen  
3        smart  
4         ford  
5     chrysler  
7   volkswagen  
12       smart  
13        audi  
14     renault  

[10 rows x 21 columns]

We need to change values in vehicle_type, gearbox, fuel_type, unrepaired_damage. Also we chack columns seller, offer_type, abtest.

In [86]:
print(autos['seller'].unique())
print(autos['offer_type'].unique())
print(autos['model'].unique())
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['privat']
['Angebot']
['andere' '7er' 'golf' 'fortwo' 'focus' 'voyager' 'a3' 'clio' nan
 'scirocco' 'a4' '911' 'cooper' '5er' 'e_klasse' 'c_klasse' 'mondeo'
 'altea' 'polo' 'a1' 'a_klasse' 'cl' '3_reihe' 's_klasse' 'sandero'
 'passat' 'corsa' 'wrangler' 'a6' 'transporter' 'astra' '3er' 'v40'
 'ibiza' 'micra' '1er' 'yaris' '6_reihe' '5_reihe' 'vito' 'cordoba'
 'galaxy' '100' '2_reihe' 'octavia' 'm_klasse' 'lupo' 'primera' 'superb'
 'meriva' 'c_max' 'laguna' 'touran' 'm_reihe' 'touareg' 'colt' 'megane'
 'avensis' 'vivaro' 'x_reihe' 'ducato' 'tigra' '1_reihe' 'boxster'
 'signum' 'sharan' 'zafira' 'ka' 'fiesta' 'rav' 'a5' 'beetle' 'c_reihe'
 'phaeton' 'i_reihe' 'sl' 'insignia' 'up' 'civic' 'mx_reihe' 'sorento'
 '80' 'z_reihe' 'berlingo' 'clk' 'punto' 's_max' 'cx_reihe' 'grand'
 'swift' 'tiguan' 'sprinter' 'mii' 'viano' 'kaefer' 'picanto' 'espace'
 'one' 'bora' 'fox' 'leon' 'transit' 'tucson' 'tt' 'qashqai' 'carnival'
 'twingo' 'mustang' 'jazz' 'almera' 'corolla' 'vectra' 'v70' 'duster'
 '

we create a dictionary with translations of these words and map it to our dataframe

In [87]:
corrections = {'privat' : 'private', 'Angebot' : 'offer', 'kleinwagen' : 'small_car', 'kombi' : 'station_wagon', 
               'cabrio' : 'convertible', 'andere' : 'other', 'manuell' : 'manually', 'automatik' : 'automatic',
               'benzin' : 'petrol', 'elektro' : 'electro', 'nein' : 'no', 'ja' : 'yes'}

In [88]:
autos['seller']=autos['seller'].map(corrections)
autos['offer_type']=autos['offer_type'].map(corrections)
autos['vehicle_type']=autos['vehicle_type'].map(corrections)
autos['gearbox']=autos['gearbox'].map(corrections)
autos['fuel_type']=autos['fuel_type'].map(corrections)
autos['unrepaired_damage']=autos['unrepaired_damage'].map(corrections)

In [89]:
print(autos['seller'].unique())
print(autos['offer_type'].unique())
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['private']
['offer']
[nan 'small_car' 'station_wagon' 'convertible' 'other']
['manually' 'automatic' nan]
[nan 'petrol' 'electro' 'other']
['no' nan 'yes']


we have create one dictionary for all columns in order not to make our code cumbersome

In [90]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37148 entries, 0 to 49999
Data columns (total 21 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        37148 non-null  object
 1   name                37148 non-null  object
 2   seller              37148 non-null  object
 3   offer_type          37148 non-null  object
 4   price               37148 non-null  int32 
 5   abtest              37148 non-null  object
 6   vehicle_type        17501 non-null  object
 7   registration_year   37148 non-null  int64 
 8   gearbox             35894 non-null  object
 9   power_ps            37148 non-null  int64 
 10  model               35775 non-null  object
 11  odometer_km         37148 non-null  int32 
 12  registration_month  37148 non-null  int64 
 13  fuel_type           21322 non-null  object
 14  brand               37148 non-null  object
 15  unrepaired_damage   31995 non-null  object
 16  ad_created          37

let's analyse 'name' column, see if there are some keywords

In [91]:
names = autos['name'].unique()

In [92]:
names[:100]

array(['Peugeot_807_160_NAVTECH_ON_BOARD',
       'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik',
       'Volkswagen_Golf_1.6_United',
       'Smart_smart_fortwo_coupe_softouch/F1/Klima/Panorama',
       'Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepflegt.mit_Klimaanlage',
       'Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Stow´n_Go_Sitze_7Sitze',
       'Golf_IV_1.9_TDI_90PS',
       'Smart_smart_fortwo_coupe_softouch_pure_MHD_Panoramadach__Klima_',
       'Audi_A3_1.6_tuning',
       'Renault_Clio_3__Dynamique_1.2__16_V;_viele_Verschleissteile_neu!',
       'Corvette_C3_Coupe_T_Top_Crossfire_Injection',
       'Volkswagen_Scirocco_2_G60',
       'mazda_tribute_2.0_mit_gas_und_tuev_neu_2018',
       'Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*DPF*AGR*TÜV*',
       'Porsche_911_Carrera_4S_Cabrio', 'MINI_Cooper_S_Cabrio',
       'Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima',
       'BMW_535i_xDrive_Sport_Aut.', 'MINI_Cooper_D',
       'Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitze_

no visible patterns

We will find the most common brand - model combination

In [93]:
brand_model = autos[['s_brand', 'model']]

In [94]:
brand_model.head()

s_brand   model
0     peugeot  andere
1         bmw     7er
2  volkswagen    golf
3       smart  fortwo
4        ford   focus

In [95]:
brand_model['comb'] = brand_model['s_brand'] + ' ' + brand_model['model']

C:\Users\Yura\AppData\Local\Temp\ipykernel_2208\78513351.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  brand_model['comb'] = brand_model['s_brand'] + ' ' + brand_model['model']


In [96]:
brand_model['comb']

0         peugeot andere
1                bmw 7er
2        volkswagen golf
3           smart fortwo
4             ford focus
              ...       
49995            audi q5
49996         opel astra
49997           fiat 500
49998            audi a3
49999        opel vectra
Name: comb, Length: 37148, dtype: object

In [97]:
brand_model['comb'].value_counts()

volkswagen golf      2918
bmw 3er              2285
volkswagen passat    1152
audi a4              1092
bmw 5er              1065
                     ... 
rover discovery         1
audi 200                1
ford b_max              1
lada samara             1
daewoo lanos            1
Name: comb, Length: 286, dtype: int64

volkswagen golf is the most popular model combination

we will split odometer_km into groups to see if average prices follow any patterns based on mileage

In [98]:
autos['odometer_km'].describe()

count     37148.000000
mean     122631.904813
std       40786.866626
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

add new column 'odometer_group' and split odometer_km into bins for 20000 km

In [113]:
autos.loc[autos['odometer_km'] < 20000, 'odometer_group'] = 'less than 20k'
autos.loc[(autos['odometer_km'] >= 20000) & (autos['odometer_km'] < 40000), 'odometer_group'] = 'less than 40k'
autos.loc[(autos['odometer_km'] >= 40000) & (autos['odometer_km'] < 60000), 'odometer_group'] = 'less than 60k'
autos.loc[(autos['odometer_km'] >= 60000) & (autos['odometer_km'] < 80000), 'odometer_group'] = 'less than 80k'
autos.loc[(autos['odometer_km'] >= 80000) & (autos['odometer_km'] < 100000), 'odometer_group'] = 'less than 100k'
autos.loc[(autos['odometer_km'] >= 100000) & (autos['odometer_km'] < 120000), 'odometer_group'] = 'less than 120k'
autos.loc[(autos['odometer_km'] >= 1200000) & (autos['odometer_km'] < 1400000), 'odometer_group'] = 'less than 140k'
autos.loc[autos['odometer_km'] >= 140000, 'odometer_group'] = 'greater than 140k'

In [114]:
autos.head()

date_crawled                                               name   seller  \
0   2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1   2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2   2016-03-26                         Volkswagen_Golf_1.6_United  private   
3   2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4   2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest   vehicle_type  registration_year    gearbox  \
0      offer   5000  control            NaN               2004   manually   
1      offer   8500  control            NaN               1997  automatic   
2      offer   8990     test            NaN               2009   manually   
3      offer   4350  control      small_car               2007  automatic   
4      offer   1350     test  station_wagon               2003   manually   

   power_ps  ... registration_month  fuel_type       brand unrepaired_damage  \
0       158  ...                  3        NaN     peugeot                no   
1       286  ...                  6     petrol         bmw                no   
2       102  ...                  7     petrol  volkswagen                no   
3        71  ...                  6     petrol       smart                no   
4         0  ...                  7     petrol        ford                no   

   ad_created nr_of_pictures postal_code   last_seen     s_brand  \
0  2016-03-26              0       79588  2016-04-06     peugeot   
1  2016-04-04              0       71034  2016-04-06         bmw   
2  2016-03-26              0       35394  2016-04-06  volkswagen   
3  2016-03-12              0       33729  2016-03-15       smart   
4  2016-04-01              0       39218  2016-04-01        ford   

      odometer_group  
0  greater than 140k  
1  greater than 140k  
2      less than 80k  
3      less than 80k  
4  greater than 140k  

[5 rows x 22 columns]

In [115]:
autos.groupby('odometer_group').mean('price')

price  registration_year    power_ps    odometer_km  \
odometer_group                                                                  
greater than 140k   4959.576718        2002.475756  129.774300  150000.000000   
less than 100k      9733.069652        2005.898010  130.040867   85394.456290   
less than 120k      9150.636665        2004.711761  126.793484  100000.000000   
less than 140k      7358.021078        2004.708982  122.861796  125000.000000   
less than 20k      23005.822222        2004.262222  126.682963    6555.555556   
less than 40k      18682.878031        2009.973609  146.136947   25192.582026   
less than 60k      15685.215393        2009.129810  129.565767   45554.279150   
less than 80k      12222.933030        2007.118907  133.373576   65125.284738   

                   registration_month  nr_of_pictures   postal_code  
odometer_group                                                       
greater than 140k            6.016342             0.0  51201.700112  
less than 100k               6.163468             0.0  52501.473703  
less than 120k               6.102154             0.0  53051.632800  
less than 140k               6.101078             0.0  52576.047665  
less than 20k                5.195556             0.0  50879.502222  
less than 40k                6.064194             0.0  53647.481455  
less than 60k                6.047674             0.0  53076.117174  
less than 80k                6.281093             0.0  53142.868337

there is a visible correlation between mileage and average price, with growing mileage price gets lower